### DISCUSSION TITLE
allowing HQL formulas in @Formula, @ColumnDefault, @Check, etc

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#4528] [Design Proposals]
Something we've discussed before, but that isn't captured here, is the ability to write HQL expressions in annotations, thus gaining portability across database platforms.

The trickiest part of this is deciding *how* to identify that the expression is HQL rather than raw SQL.

One solution would be to have some sort of "quoting" syntax, something like 

```java
@Formula("{concat(x,y)}")` 
```

or `@Formula("$concat(x,y)")` or whatever.

An alternative would be to have a new member of `@Column`, so 

```java
@Formula("concat(x,y)", dialect=HQL.class)
```

or whatever. The advantage I see to this approach is that we could also let you write things like 

```java
@Formula("x||y", dialect=PostgreSQLDialect.class) 
@Formula("concat(x,y)", dialect=MySQLDialect.class)
```

that is, write the same expression in multiple dialects of SQL.

**Comment by gavinking** [comment]
I realized that, because of the persister lifecycle, supporting HQL is quite a lot harder than it looks.

Instead I propose a semi-generic way to specify dialect-specific native SQL.

See https://github.com/hibernate/hibernate-orm/pull/4537.

**Reply by beikov** [reply]
IMO this comes down to the question: Would you still want this dialect specific native SQL if we had support for HQL expressions somehow?

I don't think these dialect specific overrides are that useful and, although I understand it's probably as good as it gets from what we can do with annotations, I don't really like the approach in the PR. I'm not saying I can propose anything better and option 3 isn't great either because all we usually want to "override" is the expression, but repeating the annotations could lead to inconsistencies for other annotation members where annotation types have more that a single SQL expression member to configure.

Overall, I think we don't need this if we simply add support for HQL expressions. WDYT?

**Reply by gavinking** [reply]
> Would you still want this dialect specific native SQL if we had support for HQL expressions somehow?

Maybe.

But after looking at what's involved on the implementation side, it's clear we're not getting HQL expressions any time soon.

The whole lifecycle is wrong.

So I think that's a sort-of academic objection.

> I don't think these dialect specific overrides are that useful

Well, that's obviously wrong, because the only reason this even came up was the guys on Zulip who were trying to do fancy stuff with DDL and just couldn't.  Perhaps it's not that useful _to you_, but the only reason I even came to look at this was because of other people ran into something they just had no way to solve.

> I don't really like the approach in the PR.

No, nor do I. It's not elegant at all. But I don't know a better way with annotations. You folks are very welcome to try and come up with something, though I suspect you won't think of anything terribly much better.

> Overall, I think we don't need this if we simply add support for HQL expressions.

Realistically: that's unlikely to happen.

**Reply by gavinking** [reply]
>  it's clear we're not getting HQL expressions any time soon.

Well, let me walk that back. In fact, it's got to be possible because we do it for JPAs `@OrderBy` annotation, in `OrderByFragmentTranslator`, so there is some existing infrastructure for this (the `PostInitCallback`, which I didn't know about).

I would have to did quite a bit deeper to find out if this is a viable way to implement *all* the things here, but at least it's a starting point. (Don't have time for it before the H6 release, however.)

> No, nor do I. It's not elegant at all. But I don't know a better way with annotations. 

So, since no-one is exactly thrilled with the `@DialectOverride` annotation, I came up with an alternative. How about if the syntax was something like:

```java
@ForDialect(dialect=DB2Dialect.class, formula=@Formula("..."), check=@Check="...")
@ForDialect(dialect=PostgreSQLDialect.class, formula=@Formula("..."), check=@Check="...")
```

That's no more verbose than the approach in the PR, and avoids dirtying up the existing annotations with `overrides` members. Actually I think it's quite a bit better.

**Reply by gavinking** [reply]
>  I came up with an alternative

Oh, wait, that's impossible, or I would have already gone down that path. The problem of course, is that I can't have a null annotation. Duh. I'm stupid, sorry for the noise.

**Reply by gavinking** [reply]
Well, um, I found a way to write a null annotation:

https://github.com/hibernate/hibernate-orm/pull/4607/files#diff-fae720d09335fc16d72b9c3224d6306d42b35e1d00074920e26f57b09e1a3b38

It's, um, ugly, but it actually works rather well. And makes the rest of the implementation of this really super-simple and low-impact.

Thoughts?

**Reply by gavinking** [reply]
A rather nice point in favor of this approach is you can use it to override JPA annotations like `@Column`.

**Reply by sebersole** [reply]
> But after looking at what's involved on the implementation side, it's clear we're not getting HQL expressions any time soon.
> 
> The whole lifecycle is wrong.

Chicken-egg.  We need the model to interpret HQL so its going to be hard to interpret HQL without the model.  The only option I can see for that is to post process the model afterwards for these overrides

**Reply by gavinking** [reply]
>  its going to be hard to interpret HQL without the model

Well that's what I thought initially. But then I noticed that we already do it for JPA's `@OrderBy`, which takes a HQL fragment. So there's at least *some* hope.

**Reply by sebersole** [reply]
> Well, let me walk that back. In fact, it's got to be possible because we do it for JPAs `@OrderBy` annotation

There we simply stash the OrderBy string and process it later (akin to what I said about post-process).  There it is easier though as it does not affect the overall structure of the model.

Though thinking more... I guess this would not affect the structure either.  Given it applies to a Column, and we really only care about the logical name at that point, we could stash these overrides on the `AnnotatedColumn` for processing later

**Reply by gavinking** [reply]
Yeah, I mean, some of the things I care most about here really only affect DDL generation.

**Reply by gavinking** [reply]
So let's just say we *can* implement this. (HQL in annotations I mean.) 

Back to the original question posed above: how would we indicate the language?

We don't want to need two annotations like with JPA `@OrderBy` vs Hibernate `@OrderBy`.

**Reply by beikov** [reply]
How about a prefix in the string e.g. `@Formula("hql: concat(x, y)")`? I can work on this for e.g. 6.1 if we agree on the syntax.

**Reply by gavinking** [reply]
@beikov I'm fine with that, at least for a prototype, and then we could think/argue about the precise syntax.

(FTR: I don't love it, but I sure as hell don't have a better suggestion.)

**Comment by sebersole** [comment]
Overall I am not a fan of these uber-annotations (one annotation to rule them all if you will).  According to this PR, all of these are perfectly legal:

```
@Entity
@ForDialect( orderBy=@OrderBy( ... ), .... )
@ForDialect( columnDefault=@ColumnDefault( ... ), .... )
@ForDialect( generatedColumn=@GeneratedColumn( ... ), .... )
@ForDialect( formula=@Formula( ... ), .... )
...
class TheEntity {
}
```

For whatever reason, y'all see this differently than I do whenever we discuss annotation design, but to me it would make much more sense to break these out.  At the very least into "valid usage targets".

**Reply by gavinking** [reply]
Sure, I don't love it either. But I don't want to add a growing set of `@OrderByForDialect`, `@FormulaForDialect`, `@CheckForDialect`.... annotations to `hibernate.annotations`.

Now, well, perhaps we could make it less-bad by making them inner annotations, something like:

```java
@Check(constraint="...")
@Check.ForDialect(dialect=OracleDialect.class, check=@Check(constraint="..."))
```

Now, that would work,  but it cuts against the idea that I wanted to reduce the impact on the existing annotations.

**Comment by sebersole** [comment]
I understand the concern about the explosion of annotations.

However, the lack of direction for the user as defined (my example) is simply not acceptable.

**Reply by gavinking** [reply]
I'm not sure I understand quite what you mean. Is your concern that someone could put `@ForDialect(orderBy=@OrderBy( ... ))` on a class instead of a field?

**Reply by sebersole** [reply]
Exactly, like in the example I gave.

**Reply by sebersole** [reply]
Personally I'd do the separate annotations, but I get that I am consistently in the minority on this :)

But I just can't accept this as defined when it allows the user to map the nonsense I did above.  Annotations should help guide them, ya know?

**Reply by sebersole** [reply]
I really do get the "explosion" aspect.  

The inner classes are not a bad idea.  Like them more the more I think about them.

**Reply by gavinking** [reply]
OK, I understand, let me think some more then.

**Reply by gavinking** [reply]
@sebersole what about:

```java
@Check(constraint="...")
@ForDialect.OverrideCheck(dialect=OracleDialect.class, check=@Check(constraint="..."))
```

Is that better or worse than:

```java
@Check(constraint="...")
@Check.ForDialect(dialect=OracleDialect.class, check=@Check(constraint="..."))
```

It's not clear to me.

**Comment by sebersole** [comment]
You mean in terms of name and placement?  One of them you added overriding for (`@DiscriminatorColumn`) is actually a JPA annotation so we'd not be able to add a nested annotation there.  So for consistency, I guess I'd vote `@ForDialect`.

Not a huge fan of the name `ForDialect`.  How about `DialectOverride` instead.

```
@Check(constraint="...")
@DialectOverride.Check(dialect=OracleDialect.class, check=@Check(constraint="..."))
```

And maybe version (major, minor, micro) makes some sense here?

**Reply by sebersole** [reply]
> And maybe version (major, minor, micro) makes some sense here?

```
@Check(constraint="...")
@DialectOverride.Check(dialect=OracleDialect.class, majorVersion=11, check=@Check(constraint="..."))
```

Just a spitball.

**Reply by gavinking** [reply]
> How about `DialectOverride` instead.

That's fine, it's what I called it in the first approach I took.

**Reply by gavinking** [reply]
> You mean in terms of name and placement?  One of them you added overriding for (`@DiscriminatorColumn`) is actually a JPA annotation so we'd not be able to add a nested annotation there. 

Yeah, sure, that's indeed the advantage of that approach. So let's go with that.

**Reply by gavinking** [reply]
>     @Check(constraint="...")
>     @DialectOverride.Check(dialect=OracleDialect.class, majorVersion=11, check=@Check(constraint="..."))

LGTM, i will implement that.

**Reply by gavinking** [reply]
OK, done: https://github.com/hibernate/hibernate-orm/pull/4607

**Reply by gavinking** [reply]
So now this looks like:

```java
@Check(constraints = "code > 0")
@DialectOverride.Check(dialect = H2Dialect.class,
		sameOrAfter = @DialectOverride.Version(major = 1),
		override = @Check(constraints = "code >= 1"))
```

Are there any more objections/feedback on this, or is it cool if I go ahead and start working on tests?

**Comment by sebersole** [comment]
Oh, I really like the version selection. 

It looks good to me

**Reply by gavinking** [reply]
> Oh, I really like the version selection.

Do you think it might be better to split that annotation out of there so it can be reused by other annotations in the future?

**Reply by gavinking** [reply]
The `@Version` annotation I mean.

**Comment by sebersole** [comment]
I'd say leave it as it is until we find such a case.  We can always add a new top-level `@DialectVersion` later if the need arises and support both on `@DialectOverride`

**Reply by gavinking** [reply]
OK, fine. It's true that I don't have a motivating use for it.

**Reply by gavinking** [reply]
Anyway I'll mark this shit `@Incubating`.
